Add some more tutorial-style documentation (#850)
Materials-Consortia/optimade-python-tools: docs/getting_started/filtering.md

# Filter parsing and transforming

> One of the aims of this package is to integrate with existing databases and APIs, and as such your particular back-end may not have a supported filter transformer.

Reviewer: ```suggestion
One of the aims of this package is to integrate with existing databases and APIs, and as such your particular backend may not have a supported filter transformer.
```

This guide will briefly outline how to parse OPTIMADE filter strings into database or API-specific queries.

## Parsing OPTIMADE filter strings

> 

Reviewer 2: I think part of the text below would fit better here.

```suggestion
The `optimade.filterparser.LarkParser` will take an OPTIMADE filter string, supplied by the user, and parse it into a `lark.Tree`.

```

```

```python
print(tree.pretty())
```

```shell
filter
  expression
    expression_clause
      expression_phrase
        comparison
          property_first_comparison
            property	nelements
            value_op_rhs
              <
              value
                number	3
```

```python
tree = p.parse('_mp_bandgap > 5.0 AND _cod_molecular_weight < 350')
print(tree.pretty())
```

```shell
filter
  expression
    expression_clause
      expression_phrase
        comparison
          property_first_comparison
            property	_mp_bandgap
            value_op_rhs
              >
              value
                number	5.0
      expression_phrase
        comparison
          property_first_comparison
            property	_cod_molecular_weight
            value_op_rhs
              <
              value
                number	350
```

## Flow for parsing a user-supplied filter and converting to a backend query

> `optimade.filterparser.LarkParser` will take user input to generate a `lark.Tree` and feed that to a `lark.Transformer`.

Reviewer 2: ```suggestion
After the `optimade.filterparser.LarkParser` has turned the filter string into a `lark.Tree`, it is fed to a `lark.Transformer` which transforms the 'lark.Tree' into a backend specific querry.
```

For example, `optimade.filtertransformers.mongo.MongoTransformer` will turn the tree into something useful for a MongoDB backend:

```python
# Example: Converting to MongoDB Query Syntax
from optimade.filtertransformers.mongo import MongoTransformer

transformer = MongoTransformer()

tree = p.parse('_mp_bandgap > 5.0 AND _cod_molecular_weight < 350')
query = transformer.transform(tree)
print(query)
```

```json
{
    "$and": [
        {"_mp_bandgap": {"$gt": 5.0}},
        {"_cod_molecular_weight": {"$lt": 350.0}}
    ]
}
```

## Developing new filter transformers

In order to support a new backend, you will need to create a new filter transformer that inherits from the [`BaseTransformer`][optimade.filtertransformers.base_transformer.BaseTransformer].
This transformer will need to override the methods that match the particular grammatical constructs in the Lark grammar in order to construct a query.

> Two examples can be found for MongoDB ([`MongoTransformer`][optimade.filtertransformers.mongo.MongoTransformer]) and Elasticsearch ([`ElasticTransformer`][optimade.filtertransformers.elasticsearch.ElasticTransformer]).

Reviewer 2: When I first read this sentence, I interpreted it as if there were two transformers for mongo DB.
Perhaps this is clearer.
```suggestion
Two examples can be found within 'optimade-python-tools', one for MongoDB ([`MongoTransformer`][optimade.filtertransformers.mongo.MongoTransformer]) and one for Elasticsearch ([`ElasticTransformer`][optimade.filtertransformers.elasticsearch.ElasticTransformer]).
```

In some cases, you may also need to write a new [`EntryCollection`][optimade.server.entry_collections.entry_collections.EntryCollection] that receives the transformed filter as an argument to its private `._run_db_query()` method.

> This class handles the connections to the underlying database, formatting of the response in an OPTIMADE format, and other API features such as sorting and pagination.

Author: ```suggestion
This class handles the connections to the underlying database, formatting of the response, and other API features such as sorting and pagination.
```

> Again, the examples for MongoDB ([`MongoCollection`][optimade.server.entry_collections.mongo.MongoCollection]) and Elasticsearch ([`ElasticCollection`][optimade.server.entry_collections.elasticsearch.ElasticCollection]) should be helpful.

Reviewer: ```suggestion
Again, this class can be a sub-class of [`EntryCollection`][optimade.server.entry_collections.entry_collections.EntryCollection], where the examples for MongoDB ([`MongoCollection`][optimade.server.entry_collections.mongo.MongoCollection]) and Elasticsearch ([`ElasticCollection`][optimade.server.entry_collections.elasticsearch.ElasticCollection]) should be helpful.
```

Author: Reworded the first sentence in this paragraph to the same effect
